docs/dev_guide.md

```diff
@@ -0,0 +1,73 @@
+# Developer's Guide
+
+Developers of OpenTurbine should understand the workflows described here.
+Adherence to these processes ensures high quality code with minimal maintenance.
+
+## Version control
+
+OpenTurbine uses git with a basic version of git-flow. The `main` branch is
```

Comment:
How will we determine when code should be merged into main?  If we ensure that all tests always pass and our test coverage is adequate, then a more stable main branch is possibly redundant.  To support official releases, we can use specific release branches, which will offer stability and possibly V&V qualification opportunities to people who need it.

Reply:
In concept, I'm ok with this approach. However, I think it relies on adhering to strict gate keeping. Pull requests must be thoroughly reviewed and well tested prior to merging. This means someone or a group of people will have to assume the role of enforcing the rules even in the face of deadlines and other work piling up. In the past, I've found that code is considered complete in order to meet a deadline (i.e. milestone) rather than when it is fully implemented and verified. The ultimate question may be is the team willing to miss a milestone to enforce code quality? If so, then merging to `main` sounds good to me and certainly simplifies the git workflow. If not, building in the redundancy of a `develop` branch can add a buffer to catch less than good code.

Reply:
There are valid reasons for having a main branch that developers don't push directly to, such as if we have nightly testing that covers more builds or runs verification problems that developers can't reasonably cover in pre-push testing.
The discussion about milestone vs good code is a larger discussion, maybe we can discuss more at the in-person mtg on the 25th. I would argue a couple of points: 1) let's not report a milestone using code that isn't fully implemented and verified. milestone calculations should be repeatable and correct. 2) I argue that good development practices such as TDD should make us go faster overall, not slower. So we shouldn't need to choose between meeting a milestone or writing good code. :-) Of course I realize there are devils in the real-world details and good intentions can melt down sometimes...